{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import toolbox  # my own little package I made to help with curation work\n",
    "from datetime import datetime\n",
    "\n",
    "today = datetime.now().strftime(\"%d%b%Y\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Dataset Creation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Rows removed: 375598         \n",
      "Percentage of original dataframe 71.62%\n"
     ]
    }
   ],
   "source": [
    "csv_path = \"../data/cleaned-data-20231214.csv\"\n",
    "df = pd.read_csv(csv_path, low_memory=False)\n",
    "\n",
    "# Feature selection\n",
    "df = df.loc[\n",
    "    :,\n",
    "    [\n",
    "        \"ID\",\n",
    "        \"TYPE\",\n",
    "        \"TABLE\",\n",
    "        \"NAME\",\n",
    "        \"FILEFORMAT\",\n",
    "        \"STUDY\",\n",
    "        \"ASSAY\",\n",
    "        \"DATATYPE\",\n",
    "        \"DATASUBTYPE\",\n",
    "        \"RESOURCETYPE\",\n",
    "    ],\n",
    "]\n",
    "\n",
    "# Explode list columns into rows\n",
    "df[\"ASSAY\"] = df[\"ASSAY\"].str.split(\",\")\n",
    "df = df.explode(\"ASSAY\")\n",
    "\n",
    "df.loc[~df.index.isin(df_full.index),].to_csv(\n",
    "    f\"../data/testing-dataset-withNulls-{today}.csv\"\n",
    ")\n",
    "\n",
    "og_shape = df.shape\n",
    "# print(\"Original dataset info\")\n",
    "# df.info()\n",
    "\n",
    "# focusing on file annotations first\n",
    "df = df[df[\"TYPE\"] != \"folder\"]\n",
    "\n",
    "# drop any missing values to develop training/test sets\n",
    "df = df.dropna(how=\"any\")\n",
    "new_shape = df.shape\n",
    "\n",
    "# print(\"-\" * 50)\n",
    "# print(\"New dataset info\")\n",
    "# df_full.info()\n",
    "\n",
    "# print(\"-\" * 50)\n",
    "print(\n",
    "    f\"Rows removed: {(np.array(og_shape) - np.array(new_shape))[0]} \\\n",
    "        \\nPercentage of original dataframe {round(((np.array(og_shape) - np.array(new_shape))[0]/np.array(og_shape))[0] * 100,2)}%\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# split dataset into train, validation, test sets\n",
    "training_percent = 0.6\n",
    "validation_percent = training_percent + 0.2\n",
    "# test set is remaining amount\n",
    "\n",
    "train, validate, test = np.split(\n",
    "    df.sample(frac=1, random_state=42),\n",
    "    [int(training_percent * len(df)), int(validation_percent * len(df))],\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>TYPE</th>\n",
       "      <th>TABLE</th>\n",
       "      <th>NAME</th>\n",
       "      <th>FILEFORMAT</th>\n",
       "      <th>STUDY</th>\n",
       "      <th>ASSAY</th>\n",
       "      <th>DATATYPE</th>\n",
       "      <th>DATASUBTYPE</th>\n",
       "      <th>RESOURCETYPE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>166907</th>\n",
       "      <td>syn30142960</td>\n",
       "      <td>file</td>\n",
       "      <td>SAGE.PORTAL_RAW.AD</td>\n",
       "      <td>29487-DLPFC.recalibrated.haplotypeCalls.annota...</td>\n",
       "      <td>TBI</td>\n",
       "      <td>AMP-AD_DiverseCohorts</td>\n",
       "      <td>whole genome sequencing assay</td>\n",
       "      <td>genomicVariants</td>\n",
       "      <td>SNPvariants</td>\n",
       "      <td>experimentalData</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152505</th>\n",
       "      <td>syn27271731</td>\n",
       "      <td>file</td>\n",
       "      <td>SAGE.PORTAL_RAW.AD</td>\n",
       "      <td>ROSMAP.SM-CJFO3.BreakSeq.vcf.gz</td>\n",
       "      <td>GZ</td>\n",
       "      <td>SV_xQTL</td>\n",
       "      <td>whole genome sequencing assay</td>\n",
       "      <td>genomicVariants</td>\n",
       "      <td>processed</td>\n",
       "      <td>analysis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11827</th>\n",
       "      <td>syn7385610</td>\n",
       "      <td>file</td>\n",
       "      <td>SAGE.PORTAL_RAW.AD</td>\n",
       "      <td>5815381006_R06C01_3-Swath1_Grn.jpg</td>\n",
       "      <td>JPG</td>\n",
       "      <td>ROSMAP</td>\n",
       "      <td>methylation array</td>\n",
       "      <td>epigenetics</td>\n",
       "      <td>raw</td>\n",
       "      <td>experimentalData</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199931</th>\n",
       "      <td>syn51720070</td>\n",
       "      <td>file</td>\n",
       "      <td>SAGE.PORTAL_RAW.AD</td>\n",
       "      <td>18207_CN.FCHMK7HDSX2_L4_ICCGCGGTT-AGCGCTAG.bam...</td>\n",
       "      <td>BAI</td>\n",
       "      <td>AMP-AD_DiverseCohorts</td>\n",
       "      <td>RNA-seq assay</td>\n",
       "      <td>geneExpression</td>\n",
       "      <td>raw</td>\n",
       "      <td>experimentalData</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110045</th>\n",
       "      <td>syn21301535</td>\n",
       "      <td>file</td>\n",
       "      <td>SAGE.PORTAL_RAW.AD</td>\n",
       "      <td>R2393217-PCC.final.bam</td>\n",
       "      <td>BAM</td>\n",
       "      <td>ROSMAP</td>\n",
       "      <td>RNA-seq assay</td>\n",
       "      <td>geneExpression</td>\n",
       "      <td>raw</td>\n",
       "      <td>experimentalData</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149419</th>\n",
       "      <td>syn27268633</td>\n",
       "      <td>file</td>\n",
       "      <td>SAGE.PORTAL_RAW.AD</td>\n",
       "      <td>MSBB.71737.BreakSeq.vcf.gz</td>\n",
       "      <td>GZ</td>\n",
       "      <td>SV_xQTL</td>\n",
       "      <td>whole genome sequencing assay</td>\n",
       "      <td>genomicVariants</td>\n",
       "      <td>processed</td>\n",
       "      <td>analysis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>187719</th>\n",
       "      <td>syn47443979</td>\n",
       "      <td>file</td>\n",
       "      <td>SAGE.PORTAL_RAW.AD</td>\n",
       "      <td>11323_TCX.FCHMLG2DSX2_L2_IATGAGGCC-GTTAATTG.bam</td>\n",
       "      <td>BAM</td>\n",
       "      <td>AMP-AD_DiverseCohorts</td>\n",
       "      <td>RNA-seq assay</td>\n",
       "      <td>geneExpression</td>\n",
       "      <td>raw</td>\n",
       "      <td>experimentalData</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>519118</th>\n",
       "      <td>syn23541635</td>\n",
       "      <td>file</td>\n",
       "      <td>SAGE.PORTAL_RAW.PSYCHENCODE</td>\n",
       "      <td>LIBD313_R2.fastq.gz</td>\n",
       "      <td>GZ</td>\n",
       "      <td>LIBD-WGBS-SCZD</td>\n",
       "      <td>bisulfite sequencing assay</td>\n",
       "      <td>epigenetics</td>\n",
       "      <td>processed</td>\n",
       "      <td>experimentalData</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173048</th>\n",
       "      <td>syn31114639</td>\n",
       "      <td>file</td>\n",
       "      <td>SAGE.PORTAL_RAW.AD</td>\n",
       "      <td>SQ_MX2024-9_S01_L003_R1_001.fastq.gz</td>\n",
       "      <td>GZ</td>\n",
       "      <td>SEA-AD</td>\n",
       "      <td>10xmultiome</td>\n",
       "      <td>geneExpression</td>\n",
       "      <td>raw</td>\n",
       "      <td>experimentalData</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156415</th>\n",
       "      <td>syn27275832</td>\n",
       "      <td>file</td>\n",
       "      <td>SAGE.PORTAL_RAW.AD</td>\n",
       "      <td>ROSMAP.SM-CTEE6.CNVnator.vcf.gz</td>\n",
       "      <td>GZ</td>\n",
       "      <td>SV_xQTL</td>\n",
       "      <td>whole genome sequencing assay</td>\n",
       "      <td>genomicVariants</td>\n",
       "      <td>processed</td>\n",
       "      <td>analysis</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>89283 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 ID  TYPE                        TABLE  \\\n",
       "166907  syn30142960  file           SAGE.PORTAL_RAW.AD   \n",
       "152505  syn27271731  file           SAGE.PORTAL_RAW.AD   \n",
       "11827    syn7385610  file           SAGE.PORTAL_RAW.AD   \n",
       "199931  syn51720070  file           SAGE.PORTAL_RAW.AD   \n",
       "110045  syn21301535  file           SAGE.PORTAL_RAW.AD   \n",
       "...             ...   ...                          ...   \n",
       "149419  syn27268633  file           SAGE.PORTAL_RAW.AD   \n",
       "187719  syn47443979  file           SAGE.PORTAL_RAW.AD   \n",
       "519118  syn23541635  file  SAGE.PORTAL_RAW.PSYCHENCODE   \n",
       "173048  syn31114639  file           SAGE.PORTAL_RAW.AD   \n",
       "156415  syn27275832  file           SAGE.PORTAL_RAW.AD   \n",
       "\n",
       "                                                     NAME FILEFORMAT  \\\n",
       "166907  29487-DLPFC.recalibrated.haplotypeCalls.annota...        TBI   \n",
       "152505                    ROSMAP.SM-CJFO3.BreakSeq.vcf.gz         GZ   \n",
       "11827                  5815381006_R06C01_3-Swath1_Grn.jpg        JPG   \n",
       "199931  18207_CN.FCHMK7HDSX2_L4_ICCGCGGTT-AGCGCTAG.bam...        BAI   \n",
       "110045                             R2393217-PCC.final.bam        BAM   \n",
       "...                                                   ...        ...   \n",
       "149419                         MSBB.71737.BreakSeq.vcf.gz         GZ   \n",
       "187719    11323_TCX.FCHMLG2DSX2_L2_IATGAGGCC-GTTAATTG.bam        BAM   \n",
       "519118                                LIBD313_R2.fastq.gz         GZ   \n",
       "173048               SQ_MX2024-9_S01_L003_R1_001.fastq.gz         GZ   \n",
       "156415                    ROSMAP.SM-CTEE6.CNVnator.vcf.gz         GZ   \n",
       "\n",
       "                        STUDY                          ASSAY         DATATYPE  \\\n",
       "166907  AMP-AD_DiverseCohorts  whole genome sequencing assay  genomicVariants   \n",
       "152505                SV_xQTL  whole genome sequencing assay  genomicVariants   \n",
       "11827                  ROSMAP              methylation array      epigenetics   \n",
       "199931  AMP-AD_DiverseCohorts                  RNA-seq assay   geneExpression   \n",
       "110045                 ROSMAP                  RNA-seq assay   geneExpression   \n",
       "...                       ...                            ...              ...   \n",
       "149419                SV_xQTL  whole genome sequencing assay  genomicVariants   \n",
       "187719  AMP-AD_DiverseCohorts                  RNA-seq assay   geneExpression   \n",
       "519118         LIBD-WGBS-SCZD     bisulfite sequencing assay      epigenetics   \n",
       "173048                 SEA-AD                    10xmultiome   geneExpression   \n",
       "156415                SV_xQTL  whole genome sequencing assay  genomicVariants   \n",
       "\n",
       "        DATASUBTYPE      RESOURCETYPE  \n",
       "166907  SNPvariants  experimentalData  \n",
       "152505    processed          analysis  \n",
       "11827           raw  experimentalData  \n",
       "199931          raw  experimentalData  \n",
       "110045          raw  experimentalData  \n",
       "...             ...               ...  \n",
       "149419    processed          analysis  \n",
       "187719          raw  experimentalData  \n",
       "519118    processed  experimentalData  \n",
       "173048          raw  experimentalData  \n",
       "156415    processed          analysis  \n",
       "\n",
       "[89283 rows x 10 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Feature Engineering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# tokenizer"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "portalanalytics-V3tNXbPg-py3.10",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
