{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from dotenv import dotenv_values\n",
    "import snowflake.connector\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from IPython.display import display, Markdown, Latex\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['SAGE.PORTAL_RAW.AD',\n",
       " 'SAGE.PORTAL_RAW.ELITE',\n",
       " 'SAGE.PORTAL_RAW.GENIE',\n",
       " 'SAGE.PORTAL_RAW.HTAN',\n",
       " 'SAGE.PORTAL_RAW.NF',\n",
       " 'SAGE.PORTAL_RAW.PSYCHENCODE']"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# there is probably a better way to do this. But this works for now\n",
    "data_tables = ['AD','ELITE', 'GENIE', 'HTAN', 'NF', 'PSYCHENCODE']\n",
    "table_names = ['SAGE.PORTAL_RAW.' + d for d in data_tables]\n",
    "table_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Initiating login request with your identity provider. A browser window should have opened for you to complete the login. If you can't see it, check existing browser windows, or your OS settings. Press CTRL+C to abort and try again...\n",
      "Going to open: https://accounts.google.com/o/saml2/idp?idpid=C030aw0es&SAMLRequest=lZLRbtowGIVfJfKuEzsJWzuLgBiMNRLdGIRJ250Tm%2BDi2MF2CPTp54RG6i5aaReRIvsc%2B%2Ft9znh6qYR3ZtpwJRMQBgh4TBaKclkmYJct%2FXvgGUskJUJJloArM2A6GRtSiRrPGnuQG3ZqmLGeO0ga3G8koNESK2K4wZJUzGBb4O3scYWjAOFaK6sKJcAry%2FsOYgzT1hEOFmq4wztYW2MI27YN2jhQuoQRQgiiz9CpOsmHQX9xM72hDyEadXqncPL1C9sXLm9P8B5WfhMZ%2FJBla3%2F9Y5sBbzagzpU0TcX0lukzL9hus7oBGEdQnZ73B0H9c%2B1441FgpGr3ghxZoaq6se7MwP3BPaNQqJK7sdNFAuojp982x%2FQp3K132fwU5ZV4uvzO7%2BPlz8bWeb5Kr5l62JBl%2Fvz1MCqA92vINepyTY1pWCq7NK1bQlHsh5EfjrLwDocfMYqD%2BO7TH%2BAtXJpcEts7B2RSFKqR1gSlUqVgPZ%2BCXXQR5LSeuo%2FTZI5iRFrkOnKrCO7v1JP%2FG3wMX3tfuvbdPX%2B6WCvBi6u3VLoi9u10wiDsVzj1970Us4pwMaNUM2NcSkKodq4Zsa7SVjcMwMnt1n9LPfkL&RelayState=50803&SigAlg=http%3A%2F%2Fwww.w3.org%2F2001%2F04%2Fxmldsig-more%23rsa-sha256&Signature=Q0TLwuq%2Ft%2FjVrDleOSyBhHPDXRK27kbOe8VAb%2F3I6lVQ31U4ceZUbpTlllXykuLXpM6JwL1IhPbBRCbgUaLuulheFu3esxweXROJFvhS9QATkU0HjuOHnpT0FejhsHdIAYD3RHi6T9AHMXO0%2FaXscW0jKdg%2FtYNEZoA6snFQOTa4FF45lGvvNKTe16BE%2FU7uW1rEdNsc58shh8432v7vUMmURow1Bn141jBWkZybidCnqSj2mJyMJR0FymfpNybLwLWM2d3tMJFvNb2qDxdwbtj7%2F7yk9LkI9ZuivZaiaRGMwqU3LS6V6QUZJbvvAAoGV141xTjA2RY9dt9lsgBaBA%3D%3D to authenticate...\n"
     ]
    }
   ],
   "source": [
    "config = dotenv_values(\"../.env\")\n",
    "conn = snowflake.connector.connect(\n",
    "    user=config['USER'],\n",
    "    account=config['ACCOUNT_IDENTIFIER'],\n",
    "    authenticator=\"externalbrowser\", # FOR browser-based SSO for authentication since account uses Google account for login. It is organization-username\n",
    "    warehouse=config['WAREHOUSE'],\n",
    "    database=config['DATABASE'],\n",
    "    role=config['ROLE'],\n",
    "    login_timeout = 60,\n",
    "    network_timeout=30,\n",
    "    socket_timeout=10\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# create cursor\n",
    "cur = conn.cursor()\n",
    "\n",
    "# Try to join all the tables together\n",
    "\n",
    "sf_tables = {}\n",
    "\n",
    "for t in table_names: \n",
    "    query = f\"\"\"\n",
    "        SELECT * FROM {t}\n",
    "    \"\"\"\n",
    "\n",
    "    cur.execute(query)\n",
    "\n",
    "    # Retrieve results\n",
    "    df = pd.concat([d for d in cur.fetch_pandas_batches()])\n",
    "    df = df.reset_index(drop=True)\n",
    "    df['TABLE'] = t\n",
    "    sf_tables[t] = df\n",
    "\n",
    "cur.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "\n",
       "|ORIGINAL|COMBINED|\n",
       "|---|---|\n",
       "|(524392, 111)|(524392, 88)|\n"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "comb_df = pd.concat(sf_tables.values()).reset_index(drop=True)\n",
    "\n",
    "original_shape = comb_df.shape\n",
    "original_cols = sorted(comb_df.columns)\n",
    "comb_df = comb_df.dropna(how='all', axis =1)\n",
    "# removing empty lists and changing all nonetypes to nans\n",
    "comb_df = comb_df.replace('[]', np.nan).fillna(value=np.nan)\n",
    "# cleanup lists and values for new lines, double spaces and quotes\n",
    "comb_df = comb_df.apply(lambda x: x.str.replace('\\n|\\s+|\"', '', regex = True), axis = 1)\n",
    "# drop empty columns\n",
    "comb_df = comb_df.dropna(how='all', axis =1)\n",
    "comb_df = comb_df[sorted(comb_df.columns)]\n",
    "\n",
    "display(Markdown(f\"\"\"\n",
    "|ORIGINAL|COMBINED|\n",
    "|---|---|\n",
    "|{original_shape}|{comb_df.shape}|\n",
    "\"\"\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ACCESSTYPE</th>\n",
       "      <th>AGE</th>\n",
       "      <th>ALIGNMENTMETHOD</th>\n",
       "      <th>ANALYSISTYPE</th>\n",
       "      <th>ASSAY</th>\n",
       "      <th>ASSAYTARGET</th>\n",
       "      <th>BENEFACTORID</th>\n",
       "      <th>BRODMANNAREA</th>\n",
       "      <th>CELLTYPE</th>\n",
       "      <th>CHROMOSOME</th>\n",
       "      <th>...</th>\n",
       "      <th>STUDYID</th>\n",
       "      <th>STUDYNAME</th>\n",
       "      <th>TABLE</th>\n",
       "      <th>TERMINALDIFFERENTIATIONPOINT</th>\n",
       "      <th>TISSUE</th>\n",
       "      <th>TRANSPLANTATIONTYPE</th>\n",
       "      <th>TREATMENTTYPE</th>\n",
       "      <th>TUMORTYPE</th>\n",
       "      <th>TYPE</th>\n",
       "      <th>VERSION</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>genotypeimputation</td>\n",
       "      <td>[snpArray]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>syn5550382</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAGE.PORTAL_RAW.AD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>genotypeimputation</td>\n",
       "      <td>[snpArray]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>syn5550382</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAGE.PORTAL_RAW.AD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>genotypeimputation</td>\n",
       "      <td>[snpArray]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>syn5550382</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAGE.PORTAL_RAW.AD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>genotypeimputation</td>\n",
       "      <td>[snpArray]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>syn5550382</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAGE.PORTAL_RAW.AD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>genotypeimputation</td>\n",
       "      <td>[snpArray]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>syn5550382</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAGE.PORTAL_RAW.AD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 88 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  ACCESSTYPE  AGE ALIGNMENTMETHOD        ANALYSISTYPE       ASSAY ASSAYTARGET  \\\n",
       "0        NaN  NaN             NaN  genotypeimputation  [snpArray]         NaN   \n",
       "1        NaN  NaN             NaN  genotypeimputation  [snpArray]         NaN   \n",
       "2        NaN  NaN             NaN  genotypeimputation  [snpArray]         NaN   \n",
       "3        NaN  NaN             NaN  genotypeimputation  [snpArray]         NaN   \n",
       "4        NaN  NaN             NaN  genotypeimputation  [snpArray]         NaN   \n",
       "\n",
       "  BENEFACTORID BRODMANNAREA CELLTYPE CHROMOSOME  ... STUDYID STUDYNAME  \\\n",
       "0   syn5550382          NaN      NaN        NaN  ...     NaN       NaN   \n",
       "1   syn5550382          NaN      NaN        NaN  ...     NaN       NaN   \n",
       "2   syn5550382          NaN      NaN        NaN  ...     NaN       NaN   \n",
       "3   syn5550382          NaN      NaN        NaN  ...     NaN       NaN   \n",
       "4   syn5550382          NaN      NaN        NaN  ...     NaN       NaN   \n",
       "\n",
       "                TABLE TERMINALDIFFERENTIATIONPOINT TISSUE TRANSPLANTATIONTYPE  \\\n",
       "0  SAGE.PORTAL_RAW.AD                          NaN    NaN                 NaN   \n",
       "1  SAGE.PORTAL_RAW.AD                          NaN    NaN                 NaN   \n",
       "2  SAGE.PORTAL_RAW.AD                          NaN    NaN                 NaN   \n",
       "3  SAGE.PORTAL_RAW.AD                          NaN    NaN                 NaN   \n",
       "4  SAGE.PORTAL_RAW.AD                          NaN    NaN                 NaN   \n",
       "\n",
       "  TREATMENTTYPE TUMORTYPE TYPE VERSION  \n",
       "0           NaN       NaN  NaN     NaN  \n",
       "1           NaN       NaN  NaN     NaN  \n",
       "2           NaN       NaN  NaN     NaN  \n",
       "3           NaN       NaN  NaN     NaN  \n",
       "4           NaN       NaN  NaN     NaN  \n",
       "\n",
       "[5 rows x 88 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "comb_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "from datetime import datetime"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "comb_df.to_csv(f\"../data/portal-data-raw-{datetime.now().strftime('%Y%m%d')}.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# pull out only columns found in all tables\n",
    "cols = {}\n",
    "\n",
    "for k,v in sf_tables.items(): \n",
    "    cols[k] = set(v.columns)\n",
    "\n",
    "# u = set.intersection(*cols.values()) # HTAN does not have any additional columns at the moment. So this is not useful"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'SAGE.PORTAL_RAW.AD': {'ANALYSISTYPE',\n",
       "  'ASSAY',\n",
       "  'ASSAYTARGET',\n",
       "  'BENEFACTORID',\n",
       "  'CELLTYPE',\n",
       "  'CHROMOSOME',\n",
       "  'CONSORTIUM',\n",
       "  'CREATEDBY',\n",
       "  'CREATEDON',\n",
       "  'CURRENTVERSION',\n",
       "  'DATAFILEHANDLEID',\n",
       "  'DATAFILESIZEBYTES',\n",
       "  'DATASUBTYPE',\n",
       "  'DATATYPE',\n",
       "  'FILEFORMAT',\n",
       "  'GRANTS',\n",
       "  'GROUPS',\n",
       "  'ID',\n",
       "  'INDIVIDUALID',\n",
       "  'INDIVIDUALIDSOURCE',\n",
       "  'ISCONSORTIUMANALYSIS',\n",
       "  'ISMODELSYSTEM',\n",
       "  'ISMULTISPECIMEN',\n",
       "  'LIBRARYPREP',\n",
       "  'METABOLITETYPE',\n",
       "  'METADATATYPE',\n",
       "  'MODELSYSTEMNAME',\n",
       "  'MODELSYSTEMTYPE',\n",
       "  'MODIFIEDBY',\n",
       "  'MODIFIEDON',\n",
       "  'NAME',\n",
       "  'NUCLEICACIDSOURCE',\n",
       "  'ORGAN',\n",
       "  'PARENTID',\n",
       "  'PROJECTID',\n",
       "  'RESOURCETYPE',\n",
       "  'SEX',\n",
       "  'SPECIES',\n",
       "  'SPECIMENID',\n",
       "  'SPECIMENIDSOURCE',\n",
       "  'STUDY',\n",
       "  'TABLE',\n",
       "  'TISSUE',\n",
       "  'TREATMENTTYPE'},\n",
       " 'SAGE.PORTAL_RAW.ELITE': {'ANALYSISTYPE',\n",
       "  'ASSAY',\n",
       "  'CONSENT',\n",
       "  'CONSORTIUM',\n",
       "  'CURRENTVERSION',\n",
       "  'DATASUBTYPE',\n",
       "  'DATATYPE',\n",
       "  'ETAG',\n",
       "  'FILEFORMAT',\n",
       "  'GRANTS',\n",
       "  'ID',\n",
       "  'ISCONSORTIUMANALYSIS',\n",
       "  'ISMODELSYSTEM',\n",
       "  'ISMULTISPECIMEN',\n",
       "  'LIBRARYPREP',\n",
       "  'LIBRARYTYPE',\n",
       "  'METADATATYPE',\n",
       "  'MODELSYSTEMTYPE',\n",
       "  'NAME',\n",
       "  'PLATFORM',\n",
       "  'PROJECT',\n",
       "  'RESOURCETYPE',\n",
       "  'SPECIES',\n",
       "  'STUDY',\n",
       "  'TABLE',\n",
       "  'TYPE'},\n",
       " 'SAGE.PORTAL_RAW.GENIE': {'BENEFACTORID',\n",
       "  'COHORT',\n",
       "  'CONSORTIUM',\n",
       "  'CREATEDBY',\n",
       "  'CREATEDON',\n",
       "  'CURRENTVERSION',\n",
       "  'DATAFILEBUCKET',\n",
       "  'DATAFILECONCRETETYPE',\n",
       "  'DATAFILEHANDLEID',\n",
       "  'DATAFILEKEY',\n",
       "  'DATAFILEMD5HEX',\n",
       "  'DATAFILENAME',\n",
       "  'DATAFILESIZEBYTES',\n",
       "  'DATATYPE',\n",
       "  'DESCRIPTION',\n",
       "  'ETAG',\n",
       "  'ID',\n",
       "  'MODIFIEDBY',\n",
       "  'MODIFIEDON',\n",
       "  'NAME',\n",
       "  'PARENTID',\n",
       "  'PROJECT',\n",
       "  'PROJECTID',\n",
       "  'SPECIES',\n",
       "  'TABLE',\n",
       "  'TYPE',\n",
       "  'VERSION'},\n",
       " 'SAGE.PORTAL_RAW.HTAN': {'COMPONENT', 'ENTITYID', 'TABLE'},\n",
       " 'SAGE.PORTAL_RAW.NF': {'ACCESSTEAM',\n",
       "  'ACCESSTYPE',\n",
       "  'AGE',\n",
       "  'ASSAY',\n",
       "  'BENEFACTORID',\n",
       "  'CELLTYPE',\n",
       "  'COMPOUNDNAME',\n",
       "  'CONSORTIUM',\n",
       "  'CONTENTTYPE',\n",
       "  'CREATEDBY',\n",
       "  'CREATEDON',\n",
       "  'DATAFILESIZEBYTES',\n",
       "  'DATASUBTYPE',\n",
       "  'DATATYPE',\n",
       "  'DIAGNOSIS',\n",
       "  'ETAG',\n",
       "  'EXPERIMENTALCONDITION',\n",
       "  'FILEFORMAT',\n",
       "  'FUNDINGAGENCY',\n",
       "  'GENEPERTURBATIONMETHOD',\n",
       "  'GENEPERTURBATIONTYPE',\n",
       "  'GENEPERTURBED',\n",
       "  'ID',\n",
       "  'INDIVIDUALID',\n",
       "  'INITIATIVE',\n",
       "  'ISCELLLINE',\n",
       "  'ISMULTIINDIVIDUAL',\n",
       "  'ISMULTISPECIMEN',\n",
       "  'ISPAIREDEND',\n",
       "  'ISSTRANDED',\n",
       "  'ISXENOGRAFT',\n",
       "  'LIBRARYPREP',\n",
       "  'LIBRARYPREPARATIONMETHOD',\n",
       "  'MODELOF',\n",
       "  'MODELSYSTEMNAME',\n",
       "  'MODIFIEDBY',\n",
       "  'MODIFIEDON',\n",
       "  'NAME',\n",
       "  'NF1GENOTYPE',\n",
       "  'NF2GENOTYPE',\n",
       "  'PARENTID',\n",
       "  'PLATFORM',\n",
       "  'PROGRESSREPORTNUMBER',\n",
       "  'PROJECTID',\n",
       "  'READPAIR',\n",
       "  'RESOURCETYPE',\n",
       "  'RESOURCE_ID',\n",
       "  'SEX',\n",
       "  'SPECIES',\n",
       "  'SPECIMENID',\n",
       "  'STUDYID',\n",
       "  'STUDYNAME',\n",
       "  'TABLE',\n",
       "  'TISSUE',\n",
       "  'TRANSPLANTATIONTYPE',\n",
       "  'TUMORTYPE',\n",
       "  'TYPE'},\n",
       " 'SAGE.PORTAL_RAW.PSYCHENCODE': {'ALIGNMENTMETHOD',\n",
       "  'ASSAY',\n",
       "  'ASSAYTARGET',\n",
       "  'BENEFACTORID',\n",
       "  'BRODMANNAREA',\n",
       "  'CELLTYPE',\n",
       "  'CONTRIBUTOR',\n",
       "  'CREATEDBY',\n",
       "  'CREATEDON',\n",
       "  'CURRENTVERSION',\n",
       "  'DATAFILEHANDLEID',\n",
       "  'DATAFILESIZEBYTES',\n",
       "  'DATASUBTYPE',\n",
       "  'DATATYPE',\n",
       "  'DIAGNOSIS',\n",
       "  'ETAG',\n",
       "  'FAILEDQC',\n",
       "  'FILEFORMAT',\n",
       "  'GRANTS',\n",
       "  'ID',\n",
       "  'INDIVIDUALID',\n",
       "  'INDIVIDUALIDSOURCE',\n",
       "  'ISMULTIINDIVIDUAL',\n",
       "  'ISMULTISPECIMEN',\n",
       "  'ISSTRANDED',\n",
       "  'LIBRARYBATCH',\n",
       "  'LIBRARYPREP',\n",
       "  'LIBRARYPREPARATIONMETHOD',\n",
       "  'MEANCOVERAGE',\n",
       "  'MEANGCCONTENT',\n",
       "  'METADATATYPE',\n",
       "  'MODIFIEDBY',\n",
       "  'MODIFIEDON',\n",
       "  'NAME',\n",
       "  'NUCLEICACIDSOURCE',\n",
       "  'ORGAN',\n",
       "  'PARENTID',\n",
       "  'PASSAGE',\n",
       "  'PH',\n",
       "  'PI',\n",
       "  'PLATFORM',\n",
       "  'PMI',\n",
       "  'PROCESSING',\n",
       "  'READEND',\n",
       "  'READLENGTH',\n",
       "  'READSTRANDORIGIN',\n",
       "  'REFERENCESET',\n",
       "  'REPROGRAMMEDCELLTYPE',\n",
       "  'RESOURCETYPE',\n",
       "  'RIN',\n",
       "  'RNABATCH',\n",
       "  'RUNTYPE',\n",
       "  'SEQUENCINGBATCH',\n",
       "  'SPECIES',\n",
       "  'SPECIMENID',\n",
       "  'SPECIMENIDSOURCE',\n",
       "  'STUDY',\n",
       "  'TABLE',\n",
       "  'TERMINALDIFFERENTIATIONPOINT',\n",
       "  'TISSUE',\n",
       "  'TRANSPOSASEBATCH'}}"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cols"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "portalanalytics-V3tNXbPg-py3.10",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
